packages/https-proxy-agent/src/index.ts

```diff
@@ -46,8 +46,13 @@ export class HttpsProxyAgent<Uri extends string> extends Agent {
 	readonly proxy: URL;
 	proxyHeaders: OutgoingHttpHeaders | (() => OutgoingHttpHeaders);
 	connectOpts: net.TcpNetConnectOpts & tls.ConnectionOptions;
+	upgradeOpts: tls.ConnectionOptions;
 
-	constructor(proxy: Uri | URL, opts?: HttpsProxyAgentOptions<Uri>) {
+	constructor(
+		proxy: Uri | URL,
+		opts?: HttpsProxyAgentOptions<Uri>,
+		upgradeOpts?: tls.ConnectionOptions
```

Comment:
I would prefer to not introduce a 3rd parameter. This should be a new option on the `opts` object. Perhaps something like `opts.tlsUpgradeOpts`.

Additionally, I think this should be an option on the `AgentBase` class, since `http-proxy-agent`, `https-proxy-agent` and `socks-proxy-agent` all upgrade the socket to TLS using similar logic. We can introduce a helper function (`AgentBase#upgradeSocketToTls()` or similar) which all of those packages would invoke when appropriate.

Finally, this needs some tests.